driver/portfolios/seq_opt_fdss_2023.py

```diff
@@ -0,0 +1,79 @@
+    # bjolp
+    (206, ['--search', 'let(lmc, landmark_cost_partitioning(lm_merged([lm_rhw(), lm_hm(m=1)])), astar(lmc,lazy_evaluator=lmc))']),
+    # seq-lmcut-por
+    (105, ['--search', 'astar(operatorcounting([state_equation_constraints(), lmcut_constraints()]), pruning=limited_pruning(pruning=atom_centric_stubborn_sets(), min_required_pruning_ratio=0.2))']),
+    # potential-initial-state
+    (83, ['--search', 'astar(initial_state_potential())']),
+    # cartesian-cegar-landmarks-goals-60s-por
+    (96, ['--search', 'astar(cegar(subtasks=[landmarks(order=random), goals(order=random)], max_states=infinity, max_transitions=infinity, max_time=60), pruning=limited_pruning(pruning=atom_centric_stubborn_sets(), min_required_pruning_ratio=0.2))']),
+    # can-sys3
+    (218, ['--search', 'astar(cpdbs(patterns=systematic(3)))']),
+    # seq-lmcut-hplus-relaxed
+    (91, ['--search', 'astar(operatorcounting([state_equation_constraints(), lmcut_constraints(), delete_relaxation_constraints(use_time_vars=false, use_integer_vars=false)]))']),
+]
+
+CONFIGS_COND_EFFS = [
+    # mas-ssc-dfp-60s
+    (1137, ['--search', 'astar(merge_and_shrink(shrink_strategy=shrink_bisimulation(greedy=false), merge_strategy=merge_sccs(order_of_sccs=topological, merge_selector=score_based_filtering(scoring_functions=[goal_relevance(), dfp(), total_order(atomic_ts_order=reverse_level,product_ts_order=new_to_old,atomic_before_product=false)])), label_reduction=exact(before_shrinking=true, before_merging=false), max_states=50k, threshold_before_merge=1, main_loop_max_time=60))']),
+    # mas-ssc-sbmiasm-300s
+    (346, ['--search', 'astar(merge_and_shrink(shrink_strategy=shrink_bisimulation(greedy=false), merge_strategy=merge_sccs(order_of_sccs=topological, merge_selector=score_based_filtering(scoring_functions=[sf_miasm(shrink_strategy=shrink_bisimulation(greedy=false),max_states=50000,threshold_before_merge=1),total_order(atomic_ts_order=reverse_level,product_ts_order=new_to_old,atomic_before_product=false)])), label_reduction=exact(before_shrinking=true, before_merging=false), max_states=50k, threshold_before_merge=1, main_loop_max_time=300))']),
+    # hmax
+    (229, ['--search', 'astar(hmax())']),
+]
+
+CONFIGS_AXIOMS = [
+    (1800, ['--search', 'astar(blind())']),
+]
+
+
+def get_pddl_features(task):
+    has_axioms = False
+    has_conditional_effects = False
+    with open(task) as f:
+        in_op = False
+        for line in f:
+            line = line.strip()
+            if line == "begin_rule":
+                has_axioms = True
+
+            if line == "begin_operator":
+                in_op = True
+            elif line == "end_operator":
+                in_op = False
+            elif in_op:
+                parts = line.split()
+                if len(parts) >= 6 and all(p.lstrip('-').isdigit() for p in parts):
```

Comment:
This seems a bit fragile to me. For example, I think `1 2 3 4 5 6` would be a legal operator name that would be considered as a conditional effect here.

Reply:
True. I'm now skipping lines with operator names. Is that enough or do you have a better solution?

Reply:
I generally think it is strange to write a poor-man's SAS^+ parser inside a portfolio file. Isn't deciding whether a task has conditional effects/axioms a problem that we will face with other portfolios as well? If so, the code for it shouldn't live inside the stone soup 2023 portfolio or be replicated in all portfolios. If the format changes, we don't want to adapt multiple parsers.

Independent of where the code lives, I would prefer a full parser to something specialized that just checks for specific features. Something like
```Python
task = parse_task(sas_file)
if task.has_conditional_effects() or task.has_axioms():
    whatever()
```

The current code also hard-codes the input file as `output.sas` and the users could pick a different name for the sasfile, couldn't they?


Another option could be to go the route we did for unit cost tasks, where we have `--is-unit-cost` as a flag that is handled by the search component. With similar flags for `--has-axioms` and `--has-conditional-effects` we could leave the portfolio file static and just return a fixed configuration.

Maybe we should move this discussion to the tracker as well? Or to a Fast Downward meeting?

Reply:
I suggest going for something along the lines of the current "--if-unit-cost" option. We wanted to modernize this to an expression using the new option parser, has this already happened? If not, we can use what these portfolios needs as a new use case for designing the feature. So far we planned to keep it simple, along the lines of an "ifunitcost(config1, config2)" expression. (Perhaps with underscores: if_unit_cost.) I don't know where this discussion currently lives on the tracker, if anywhere.

Reply:
Perhaps we can do all this more easily once we have librarified the planner more. For the time being, one alternative could be a planner mode/option that just evaluates some tests (does the task have axioms etc.) and then returns an exit code signaling the result. Then the portfolio code could leverage the existing code in the planner, but still do its own conditionals.

Reply:
That sounds like we're resurrecting the old Rube Goldberg machine ;-) I'm reluctant about replacing a simple hack with a complicated hack, especially if a better solution (the Python interface) is on the horizon. I'd just like to have the simple version of this merged, since we already got an inquiry of someone who wants to use the FDSS portfolios within the latest codebase.

Reply:
What's your stance on this, Florian and Malte?

Reply:
From my side, whatever implementation we use to get to the information shouldn't live inside the portfolio script. The hard-coded `output.sas` is an indication that this is wrong place for it.

If we generally want to support portfolios that have different configurations based on things like axioms and conditional effects, then maybe we should pass this information into the portfolio script somehow. Alternatively, we could say that a portfolio is by definition a static list of timed configurations (as the other portfolios are) and introduce a new concept above them that does the switching. For example, a "DynamicPortfolio" (or a better name) could take in a task and return a portfolio (list of timed configurations).

Reply:
OK, then let's add only the satisficing portfolio. Does the rest of the patch look good to merge?

Reply:
I'm fine with adding the portfolio, I don't get the changes in the tests.

You could also add the optimal portfolios individually, e.g. add
* `seq_sat_fdss_2023.py` for the satisficing portfolio,
* `seq_opt_fdss_2023_axioms.py` for the optimal portfolio in the presence of axioms,
* `seq_opt_fdss_2023_condeffs.py` for the optimal portfolio in the presence of conditional effects, and
* `seq_opt_fdss_2023_strips.py` for the optimal portfolio for STRIPS tasks.

Reply:
I can have a look at this later, but are you sure you don't want to consider having the search component look at the task properties? I don't agree that this is any way like the old Rube Goldberg code. It's already the case in the portfolio code that we repeatedly call the search, "dry-run" features are not at all unusual, and we'd use the task parsing code we already have. Also, having a scriptable test of task properties like this would be something I could see other people using as well in their pipelines.

Here is an example diff for the search code that allows us to test whether the task has conditional effects with `--test-conditional-effects'):

```
 git diff | cat
diff --git a/src/search/command_line.cc b/src/search/command_line.cc
index 5b95ff424..64748f198 100644
--- a/src/search/command_line.cc
+++ b/src/search/command_line.cc
@@ -2,14 +2,18 @@
 
 #include "plan_manager.h"
 #include "search_algorithm.h"
+#include "task_proxy.h"
 
 #include "parser/lexical_analyzer.h"
 #include "parser/syntax_analyzer.h"
 #include "plugins/any.h"
 #include "plugins/doc_printer.h"
 #include "plugins/plugin.h"
+#include "task_utils/task_properties.h"
+#include "tasks/root_task.h"
 #include "utils/logging.h"
 #include "utils/strings.h"
+#include "utils/system.h"
 
 #include <algorithm>
 #include <sstream>
@@ -149,6 +153,13 @@ static shared_ptr<SearchAlgorithm> parse_cmd_line_aux(const vector<string> &args
             num_previously_generated_plans = parse_int_arg(arg, args[i]);
             if (num_previously_generated_plans < 0)
                 input_error("argument for --internal-previous-portfolio-plans must be positive");
+        } else if (arg == "--test-conditional-effects") {
+            TaskProxy task_proxy(*tasks::g_root_task);
+            if (task_properties::has_conditional_effects(task_proxy)) {
+                exit(static_cast<int>(utils::ExitCode::SEARCH_UNSUPPORTED));
+            } else {
+                exit(static_cast<int>(utils::ExitCode::SUCCESS));
+            }
         } else {
             input_error("unknown option " + arg);
         }
```
```

Reply:
Thanks for the suggestions, both of you. But I think having the satisficing portfolio merged is enough for now.

Is that fine for you @maltehelmert ?

Reply:
Sure, works for me. Hopefully this can "soon" be done with the Python interface.